### PR TITLE
Fix/forbidden manifest

### PR DESCRIPTION
**Pull request by pritishmishra**
Modifying the approach of finding forbidden manifest changes, as per the bosh2 manifest changes

Related to https://github.com/cloudfoundry-incubator/service-fabrik-broker/issues/11

### CODE REVIEW

File: lib/utils/index.js [hunk 1]
```diff
@@ -341,17 +341,37 @@ function hasChangesInForbiddenSections(diff) {
...
     .value();
+
+  function isJobRemoved() {
+    const jobsRegex = new RegExp('^  jobs'); // this regex is to find the position of jobs section
+    const jobsLevelRegex = new RegExp('^  [a-z]+'); // this regex is to find the position of next section at the same level as jobs
+    const jobNameRegex = new RegExp('^  [- ] name'); // this regex is to find the position of removal in job name
+    const jobStartIndex = _.findIndex(diff, element => {
+      return jobsRegex.test(element[0]);
+    });
+    const jobEndIndex = _.findIndex(diff, element => {
```
Comment (jagadish-kb) [inline lib/utils/index.js]: similarly above can be written like below. If the method has only one statement and return, then follow the below concise way.
```
const jobEndIndex = _.findIndex(diff, jobsLevelRegex.test(element[0]), jobStartIndex + 1);
```

File: lib/utils/index.js [hunk 2]
```diff
@@ -341,17 +341,37 @@ function hasChangesInForbiddenSections(diff) {
...
+      return jobsLevelRegex.test(element[0]);
+    }, jobStartIndex + 1);
+
+    const jobDiff = _.slice(diff, jobStartIndex, jobEndIndex);
+    const jobNameRemovedIndex = _.findIndex(jobDiff, element => {
```
Comment (jagadish-kb) [inline lib/utils/index.js]: same as above

File: lib/utils/index.js [hunk 3]
```diff
@@ -341,17 +341,37 @@ function hasChangesInForbiddenSections(diff) {
...
+    });
+    if (jobNameRemovedIndex !== -1) {
+      return true;
+    } // this checks if the name removal is present within the jobs section only.
+    return false;
+  }
   if (!_.isEmpty(forbiddenSections) && !_.includes(forbiddenSections, 'director_uuid')) {
-    throw new errors.Forbidden(`Automatic update not possible. Detected changes in forbidden section(s) '${forbiddenSections.join(',')}'`);
+    const forbiddenSectionsDiff = _.filter(diff, element => _.includes(element[0], 'instances') || _.includes(element[0], 'persistent_disk_type'));
+    if (!_.isEmpty(forbiddenSectionsDiff) || isJobRemoved()) {
```
Comment (jagadish-kb) [inline lib/utils/index.js]: you are not passing any args to the function `isJobRemoved()` how do you pass `diff` to it? Has this passed UT?

File: test/jobs.ServiceInstanceUpdateJob.spec.js [hunk 4]
```diff
@@ -236,10 +236,10 @@ describe('Jobs', function () {
         organization_guid: organization_guid
       });
       const diff = [
-        ['jobs:', null],
-        ['- name: blueprint_z1', null],
-        ['  instances: 2', 'removed'],
-        ['  instances: 1', 'added']
+        ['instance_groups:', null],
```
Comment (jagadish-kb) [inline test/jobs.ServiceInstanceUpdateJob.spec.js]: ADD UT for the new section of Code written. Exclusively test with all the changes ex., instance group, persitence disk, job removal.

File: lib/utils/index.js [hunk 5]
```diff
@@ -341,17 +341,37 @@ function hasChangesInForbiddenSections(diff) {
...
+      throw new errors.Forbidden(`Automatic update not possible. Detected changes in forbidden section(s) '${forbiddenSections.join(',')}'`);
```
Comment (jagadish-kb) [inline lib/utils/index.js]: change the error message here to reflect the actual changes. for ex, it should reflect the values from `forbiddenSectionsDiff` and not from `forbiddenSections`

File: lib/utils/index.js (hunk 3 above)
Comment (pritishmishra) [reply lib/utils/index.js]: `diff` is already a local variable to the function `hasChangesInForbiddenSections` and the function `isJobRemoved()` is present within this function. So, do I require to pass `diff` again?

File: test/jobs.ServiceInstanceUpdateJob.spec.js (hunk 4 above)
Comment (pritishmishra) [reply test/jobs.ServiceInstanceUpdateJob.spec.js]: ok.. will add another test case which tests the job removal. Persistence disk type is exactly the same as instances, the call flow is same too. So, adding a test case for job removal which will test the newly added functions.

File: lib/utils/index.js (hunk 5 above)
Comment (pritishmishra) [reply lib/utils/index.js]: ok.. I will change the message

File: lib/utils/index.js (hunk 2 above)
Comment (pritishmishra) [reply lib/utils/index.js]: ok

File: lib/utils/index.js (hunk 1 above)
Comment (pritishmishra) [reply lib/utils/index.js]: ok

File: lib/utils/index.js [hunk 6]
```diff
@@ -341,17 +341,37 @@ function hasChangesInForbiddenSections(diff) {
...
       'tags',
       'addons'
     ])
     .value();
+
+  function isJobRemoved() {
+    const jobsRegex = new RegExp('^  jobs'); // this regex is to find the position of jobs section
+    const jobsLevelRegex = new RegExp('^  [a-z]+'); // this regex is to find the position of next section at the same level as jobs
+    const jobNameRegex = new RegExp('^  [- ] name'); // this regex is to find the position of removal in job name
+    const jobStartIndex = _.findIndex(diff, element => {
```
[the remark this answers is not in the document]
Comment (pritishmishra) [reply lib/utils/index.js]: ok

File: lib/utils/index.js [hunk 7]
```diff
@@ -334,24 +334,37 @@ function isServiceFabrikOperationFinished(state) {
...
     ])
     .value();
+
   if (!_.isEmpty(forbiddenSections) && !_.includes(forbiddenSections, 'director_uuid')) {
-    throw new errors.Forbidden(`Automatic update not possible. Detected changes in forbidden section(s) '${forbiddenSections.join(',')}'`);
+    const forbiddenSectionsDiff = _.filter(diff, element => _.includes(element[0], 'instances') || _.includes(element[0], 'persistent_disk_type'));
+    const removedJobName = findRemovedJob();
+    if (!_.isEmpty(forbiddenSectionsDiff) || removedJobName) {
+      throw new errors.Forbidden(`Automatic update not possible. ${!_.isEmpty(forbiddenSectionsDiff)? 'Detected changes in forbidden sections:' + forbiddenSectionsDiff.join(','): `Job definition removed: ${removedJobName}`}`);
+    }
```
Comment (jagadish-kb) [inline lib/utils/index.js]: make this ${removedJobName[0]}

**Pull request merged**